src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransactionStore.cs

```diff
@@ -0,0 +1,216 @@
+                    transaction.ValuesLazyLoadingIsOn = false;
+
+                    Row<byte[], TemplateTransaction> templateRow = transaction.Select<byte[], TemplateTransaction>(TemplateTransactionTableName, hash.ToBytes());
+
+                    if (templateRow.Exists)
+                    {
+                        res = templateRow.Value;
+                    }
+                }
+
+                this.logger.LogTrace("(-):{0}", res);
+                return res;
+            });
+
+            return task;
+        }
+
+        /// <inheritdoc />
+        public Task PutAsync(TemplateTransaction templateTransaction)
+        {
+            Guard.NotNull(templateTransaction, nameof(templateTransaction));
+
+            Task task = Task.Run(() =>
+            {
+                this.logger.LogTrace("()");
+
+                using (DBreeze.Transactions.Transaction transaction = this.DBreeze.GetTransaction())
+                {
+                    this.OnInsertTemplateTransaction(transaction, templateTransaction);
+
+                    transaction.Commit();
+                }
+
+                this.logger.LogTrace("(-)");
+            });
+
+            return task;
+        }
+
+        /// <inheritdoc />
+        public Task<bool> ExistAsync(uint256 hash)
+        {
+            Guard.NotNull(hash, nameof(hash));
+
+            Task<bool> task = Task.Run(() =>
+            {
+                this.logger.LogTrace("()");
+
+                bool res = false;
+                using (DBreeze.Transactions.Transaction transaction = this.DBreeze.GetTransaction())
+                {
+                    // Lazy loading is on so we don't fetch the whole value, just the row.
+                    Row<byte[], TemplateTransaction> templateRow = transaction.Select<byte[], TemplateTransaction>(TemplateTransactionTableName, hash.ToBytes());
+
+                    if (templateRow.Exists)
+                    {
+                        res = true;
+                    }
+                }
+
+                this.logger.LogTrace("(-):{0}", res);
+                return res;
+            });
+
+            return task;
+        }
+
+        /// <inheritdoc />
+        public Task DeleteAsync(uint256 hash)
+        {
+            Guard.NotNull(hash, nameof(hash));
+
+            Task task = Task.Run(() =>
+            {
+                this.logger.LogTrace("()");
+
+                using (DBreeze.Transactions.Transaction transaction = this.DBreeze.GetTransaction())
+                {
+                    transaction.ValuesLazyLoadingIsOn = false;
+
+                    this.OnDeleteTemplateTransaction(transaction, hash);
+
+                    transaction.Commit();
+                }
+
+                this.logger.LogTrace("(-)");
+            });
+
+            return task;
+        }
+
+        protected virtual void OnInsertTemplateTransaction(DBreeze.Transactions.Transaction dbreezeTransaction, TemplateTransaction templateTransaction)
```

Comment:
'On' suggests an event or signal has happened, maybe change to `InsertTemplateTransaction`.